### PR TITLE
Warn when Using DefaultProps on Function Components

### PR DESCRIPTION
**Pull request by lunaruan**
As part of the process to deprecate defaultProps on function components (as per a larger proposal outlined [here](https://github.com/reactjs/rfcs/blob/createlement-rfc/text/0000-create-element-changes.md)), add a warning whenever someone does this.

### CODE REVIEW

File: packages/react-dom/src/__tests__/ReactFunctionComponent-test.js [hunk 1]
```diff
@@ -365,7 +386,11 @@ describe('ReactFunctionComponent', () => {
     );
   });
 
+  // TODO: deprecate default props support after we remove defaultProps
+  // from function components
```
Comment (gaearon) [inline packages/react-dom/src/__tests__/ReactFunctionComponent-test.js]: Didn't we plan to keep them for classes forever?

Comment (lunaruan) [reply packages/react-dom/src/__tests__/ReactFunctionComponent-test.js]: Oh sorry, poor phrasing. I just meant that I need to change this test at some point. Will rephrase!

File: packages/react-dom/src/__tests__/ReactFunctionComponent-test.js [hunk 2]
```diff
@@ -25,6 +26,8 @@ describe('ReactFunctionComponent', () => {
     React = require('react');
     ReactDOM = require('react-dom');
     ReactTestUtils = require('react-dom/test-utils');
+    ReactFeatureFlags = require('shared/ReactFeatureFlags');
+    ReactFeatureFlags.warnAbouDefaultPropsOnFunctionComponents = true;
```
Comment (threepointone) [inline packages/react-dom/src/__tests__/ReactFunctionComponent-test.js]: s/`warnAbouDefaultPropsOnFunctionComponents`/`warnAboutDefaultPropsOnFunctionComponents` (note the extra t at the end of `About`)

File: packages/react-dom/src/__tests__/ReactFunctionComponent-test.js [hunk 3]
```diff
@@ -25,6 +26,8 @@ describe('ReactFunctionComponent', () => {
...
   });
```
Comment (threepointone) [inline packages/react-dom/src/__tests__/ReactFunctionComponent-test.js]:
```suggestion

  afterEach(() => {
    ReactFeatureFlags.warnAboutDefaultPropsOnFunctionComponents = false;
  })
```

File: packages/react-dom/src/__tests__/ReactFunctionComponent-test.js [hunk 4]
```diff
@@ -206,6 +209,24 @@ describe('ReactFunctionComponent', () => {
...
+    }
+
+    FunctionalComponent.defaultProps = {
+      testProp: true,
+    };
+
+    expect(() =>
+      ReactTestUtils.renderIntoDocument(<FunctionalComponent />),
+    ).toWarnDev(
+      'Warning: FunctionalComponent: Function components do not support defaultProps. ' +
```
Comment (ljharb) [inline packages/react-dom/src/__tests__/ReactFunctionComponent-test.js]: *do not*, or *will not* in a future version?

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 5]
```diff
@@ -1424,6 +1427,23 @@ function validateFunctionComponentInDev(workInProgress: Fiber, Component: any) {
...
+    warnAbouDefaultPropsOnFunctionComponents &&
+    Component.defaultProps !== undefined
+  ) {
+    const componentName = getComponentName(Component) || 'Unknown';
+
+    if (!didWarnAboutDefaultPropsOnFunctionComponent[componentName]) {
+      warningWithoutStack(
+        false,
+        '%s: Function components do not support defaultProps. ' +
+          'Use Javascript default arguments instead.',
```
Comment (ljharb) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]:
```suggestion
          'Use JavaScript default arguments instead.',
```

File: packages/react-dom/src/__tests__/ReactFunctionComponent-test.js [hunk 6]
```diff
@@ -206,6 +209,24 @@ describe('ReactFunctionComponent', () => {
...
+        'Use Javascript default arguments instead.',
```
Comment (ljharb) [inline packages/react-dom/src/__tests__/ReactFunctionComponent-test.js]:
```suggestion
        'Use JavaScript default arguments instead.',
```

**Comment by acdlite** [comment]
Tests suites that modify feature flags need an `internal` suffix.

The reason is that we have a CI job that runs our unit tests against the actual npm packages. `shared/ReactFeatureFlags` doesn't exist in those packages, because it gets inlined by our build pipeline.

**Comment by acdlite** [comment]
Need to update all the feature flag files to include the new one you added. E.g. `ReactFeatureFlags.native-oss.js`. That's why the Flow and build CI jobs are failing.

File: packages/react-reconciler/src/ReactFiberBeginWork.js [hunk 7]
```diff
@@ -1424,6 +1427,23 @@ function validateFunctionComponentInDev(workInProgress: Fiber, Component: any) {
...
+    warnAboutDefaultPropsOnFunctionComponents &&
+    Component.defaultProps !== undefined
+  ) {
+    const componentName = getComponentName(Component) || 'Unknown';
+
+    if (!didWarnAboutDefaultPropsOnFunctionComponent[componentName]) {
+      warningWithoutStack(
+        false,
+        '%s: Function components will not support defaultProps. ' +
+          'Use JavaScript default arguments instead.',
```
Comment (FredyC) [inline packages/react-reconciler/src/ReactFiberBeginWork.js]: Might be worth mentioning it's ES2015 feature ... or not. 🤷‍♂

File: packages/react-dom/src/__tests__/ReactDeprecationWarnings-test.internal.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+    FunctionalComponent.defaultProps = {
+      testProp: true,
+    };
+
+    expect(() =>
+      ReactTestUtils.renderIntoDocument(<FunctionalComponent />),
+    ).toWarnDev(
+      'Warning: FunctionalComponent: Function components will not support defaultProps. ' +
+        'Use JavaScript default arguments instead.',
```
Comment (acdlite) [inline packages/react-dom/src/__tests__/ReactDeprecationWarnings-test.internal.js]: Alt suggestion for first sentence: "Support for `defaultProps` will be removed from function components in a future release."

Comment (ljharb) [reply packages/react-dom/src/__tests__/ReactDeprecationWarnings-test.internal.js]: ideally, a future *major* release

Comment (acdlite) [reply packages/react-dom/src/__tests__/ReactDeprecationWarnings-test.internal.js]: Right, I assume that was implied but should probably be specific

**Comment by Sawtaytoes** [comment]
I don't know why we're deprecating `defaultProps` on function components. I mean, I get the concept, but please look at this:

image

The second and third examples are functionally the same, but the first is different. If React renders `<MyComponent />` again, it will create a new default object for `value`.

Am I understanding this correctly?

What happens if you have `useEffect` calls like this:
```js
const [
  localValues,
  setLocalValues,
] = useState(values)

useEffect(
  () => {
    setLocalValues(
      values
    )
  },
  [values],
)

return (
  <div>
    {values.text}
  </div>
)
```

The first example would call the `useEffect` more times than you expect whereas the latter two only call it the one time (at mount).  While you may think "a few extra Virtual DOM diffs is no big deal", it will lead to many unexpected bugs in hooks because you're writing with the intention that props, which were never passed, are never changing.

I myself have run into this issue as well and kept wondering why my values kept getting overridden when `children` updated even though I wasn't passing any props.

**Comment by FredyC** [comment]
@Sawtaytoes If you are into endless debate on this topic, be sure to read through https://github.com/reactjs/rfcs/pull/107. I am sure you will find your answers there.

**Comment by gaearon** [comment]
The PR is for the implementation so any discussion here will get lost.

**Comment by Sawtaytoes** [comment]
Where should I move the discussion? The RFC place?

**Comment by gaearon** [comment]
Sure.

**Comment by uloco** [comment]
When this gets depreacted, is there still a way to set the defaultProps globally for example the react-native <Text> element? Overriding default props of existing components from react / react-native, which I don't have control of is where this is really good to have it available as an option..

**Comment by uloco** [comment]
It's exactly what I am talking about. I want the default `<Text />` element for android to always have the property `android_hyphenationFrequency` to be set to  `full`. Because I want all text elements throughout the app to have this behaviour. Should I really create a custom component for this?

Until now it works (in react 17) that you override it globally with `Text.defaultprops.android_hypenationFrequency = "full"`. This is also a common pattern for font scaling, because the defaults are just not that great here.

For me it feels like a weird pattern to have a different custom component for a behaviour that is so basic as this and needed in every single place always.

You could warn the user to not use this in components they create themselves, but being able to customize someone elses defaults would be nice imho...

**Comment by ljharb** [comment]
Yes, you really should. Don’t mutate objects you don’t own.

**Pull request merged**